proto/sql_api.proto

```diff
@@ -0,0 +1,154 @@
+  // CmdID is optionally specified for request idempotence
+  // (i.e. replay protection).
+  optional ClientCmdID cmd_id = 4 [(gogoproto.nullable) = false, (gogoproto.customname) = "CmdID"];
+  // User is the originating user. Used to lookup priority when
+  // scheduling queued operations at target node.
+  optional string user = 5 [(gogoproto.nullable) = false];
+  // UserPriority specifies priority multiple for non-transactional
+  // commands. This value should be a positive integer [1, 2^31-1).
+  // It's properly viewed as a multiple for how likely this
+  // transaction will be to prevail if a write conflict occurs.
+  // Commands with UserPriority=100 will be 100x less likely to be
+  // aborted as conflicting transactions or non-transactional commands
+  // with UserPriority=1. This value is ignored if Txn is
+  // specified. If neither this value nor Txn is specified, the value
+  // defaults to 1.
+  optional int32 user_priority = 6 [default = 1];
+  // Txn is set non-nil if a transaction is underway. To start a txn,
+  // the first request should set this field to non-nil with name and
+  // isolation level set as desired. The response will contain the
+  // fully-initialized transaction with txn ID, priority, initial
+  // timestamp, and maximum timestamp. The transaction returned in a
+  // response is feed to the next request.
+  optional Transaction txn = 7;
+  // ReadConsistency specifies the consistency for read
+  // operations. The default is CONSISTENT. This value is ignored for
+  // write operations.
+  optional ReadConsistencyType read_consistency = 8 [(gogoproto.nullable) = false];
+}
+
+// CmdResponseHeader is returned with every Cmd response.
+message CmdResponseHeader {
+  // Error is non-nil if an error occurred.
+  optional Error error = 1;
+  // Transaction is non-nil if the request specified a non-nil
+  // transaction. The transaction timestamp and/or priority may have
+  // been updated, depending on the outcome of the request.
+  optional Transaction txn = 3;
+}
+
+// A sql CmdRequest to cockroach. A transaction can consists of multiple
+// CmdRequest.
+message CmdRequest {
+  optional CmdRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional string sql = 2 [(gogoproto.nullable) = false];
+  // parameters are referred to in the above sql query using $1, $2, etc.
+  // params[n-1] -> $n
+  repeated bytes params = 3;
+}
+
+message CmdResponse {
+  optional CmdResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  // The result set.
+  repeated Result results = 2;
+}
+
+// A Result is a collection of values representing a row
+// in a result view. A column value not present in a row
+// has Nil Bytes in the value.
+message Result  {
+  repeated Value values = 1;
+}
+
+// An EndTransactioniCmdRequest specifies whether to commit or roll back an
+// extant transaction.
+message EndTransactionCmdRequest {
```

Comment:
Sounds good. We can also support SQL servers BEGIN TRANSACTION later. I'd also like the sql to be a repeated field so that a user can pass in a sequence of commands that include a COMMIT at the end if that's how they want to write it.